File: Roadmap/08 - CLASES/python/Sac-Corts.py
"""
EJERCICIO
"""
class Persona:
    def __init__(self, nombre, edad):
        self.nombre = nombre
        self.edad = edad

    def saludar(self):
        self.saludar = print(f"Hola {self.nombre} ({self.edad}), ¿cómo estás?")

isaac = Persona("Isaac", 22)
isaac.saludar()

"""
EJERCICIO EXTRA
"""

class Stack:
    def __init__(self):
        self.items = []

    def push(self, item):
        self.items.append(item)
    
    def pop(self):
        if not self.is_empty():
            return self.items.pop()
        else: 
            return None
    
    def peek(self):
        if not self.is_empty():
            return self.items[-1]
        else:
            return None
        
    def is_empty(self):
        return len(self.items) == 0
    
    def size(self):
        size = len(self.items)
        items = self.items[::-1]
        print(f"{size} | {items}")
        return size, items  
    
pila = Stack()
pila.push(10)
pila.push(20)
pila.push(30)
print(pila.is_empty())
dimension, elementos = pila.size()
print(pila.pop())
print(pila.peek())

class Queue:
    def __init__(self):
        self.items = []

    def enqueue(self, item):
        self.items.append(item)
    
    def dequeue(self):
        if not self.is_empty():
            return self.items.pop(0)
        else:
            return None
    
    def peek(self):
        if not self.is_empty():
            return self.items[0]
        else:
            return None
    
    def is_empty(self):
        return len(self.items) == 0
    
    def size(self):
        size = len(self.items)
        items = self.items
        print(f"{size} | {items}")
        return size, items
    
cola = Queue()
cola.enqueue(11)
cola.enqueue(12)
cola.enqueue(13)
print(pila.is_empty())
dimension, elementos = cola.size()
print(cola.dequeue())
print(cola.peek())

File: Roadmap/08 - CLASES/typescript/Sac-Corts.ts
class Person {
    name: string;
    age: number;
    job: string;

    constructor(name: string, age: number, job: string) {
        this.name = name;
        this.age = age;
        this.job = job;
    }

    printDetails(): void {
        console.log(`Name: ${this.name}`);
        console.log(`Age: ${this.age}`);
        console.log(`Job: ${this.job}`);
    }
}

let person1 = new Person("Isaac", 22, "Backend Developer");
person1.printDetails();

person1.name = "Yusuke";
person1.age = 14;
person1.job = "Fighter";
person1.printDetails();

// *** Extra Exercise *** //

// Stack
class Stack<T> {
    private elements: T[] = [];

    push(element: T): void {
        this.elements.push(element);
    }

    pop(): T | undefined {
        return this.elements.pop();
    }

    size(): number {
        return this.elements.length;
    }

    print(): void {
        console.log("Stack: ", this.elements);
    }
}

let stack = new Stack<number>();
stack.push(10);
stack.push(11);
console.log(stack.print());
console.log(stack.pop());
console.log(stack.size());

// Queue
class Queue<T> {
    private elements: T[] = [];

    enqueue(element: T): void {
        this.elements.push(element);
    }

    dequeue(): T | undefined {
        return this.elements.shift();
    }

    size(): number {
        return this.elements.length;
    }

    print(): void {
        console.log("Queue: ", this.elements);
    }
}

let queue = new Queue<string>();
queue.enqueue("A");
queue.enqueue("B");
console.log(queue.print());
console.log(queue.dequeue());
console.log(queue.size());
